pkg/etcdcli/etcdcli.go

```diff
@@ -68,82 +63,63 @@ func NewEtcdClient(kubeInformers v1helpers.KubeInformersForNamespaces, networkIn
+			return nil, fmt.Errorf("configmaps lister not synced")
+		}
+		if !g.networkListerSynced() {
+			return nil, fmt.Errorf("network lister not synced")
+		}
 
-	etcdEndpoints := []string{}
-	nodes, err := g.nodeLister.List(labels.Set{"node-role.kubernetes.io/master": ""}.AsSelector())
-	for _, node := range nodes {
-		internalIP, err := dnshelpers.GetEscapedPreferredInternalIPAddressForNodeName(network, node)
+		network, err := g.networkLister.Get("cluster")
 		if err != nil {
-			return nil, fmt.Errorf("failed to get internal IP for node: %w", err)
+			return nil, fmt.Errorf("failed to list cluster network: %w", err)
 		}
-		etcdEndpoints = append(etcdEndpoints, fmt.Sprintf("https://%s:2379", internalIP))
-	}
 
-	configmap, err := g.configmapsLister.ConfigMaps(operatorclient.TargetNamespace).Get("etcd-endpoints")
-	if err != nil {
-		return nil, fmt.Errorf("failed to list endpoints: %w", err)
-	}
-	if bootstrapIP, ok := configmap.Annotations[BootstrapIPAnnotationKey]; ok && bootstrapIP != "" {
-		// escape if IPv6
-		if net.ParseIP(bootstrapIP).To4() == nil {
-			bootstrapIP = "[" + bootstrapIP + "]"
+		var etcdEndpoints []string
+		nodes, err := g.nodeLister.List(labels.Set{"node-role.kubernetes.io/master": ""}.AsSelector())
+		for _, node := range nodes {
+			internalIP, err := dnshelpers.GetEscapedPreferredInternalIPAddressForNodeName(network, node)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get internal IP for node: %w", err)
+			}
+			etcdEndpoints = append(etcdEndpoints, fmt.Sprintf("https://%s:2379", internalIP))
+		}
+
+		configmap, err := g.configmapsLister.ConfigMaps(operatorclient.TargetNamespace).Get("etcd-endpoints")
+		if err != nil {
+			return nil, fmt.Errorf("failed to list endpoints: %w", err)
 		}
-		etcdEndpoints = append(etcdEndpoints, fmt.Sprintf("https://%s:2379", bootstrapIP))
-	}
-
-	g.clientLock.Lock()
-	defer g.clientLock.Unlock()
-	if reflect.DeepEqual(g.lastClientConfigKey, etcdEndpoints) && g.cachedClient != nil {
-		ctx, cancel := context.WithCancel(context.Background())
-		defer cancel()
-		// Test cached client connection by doing a request.
-		_, err = g.cachedClient.MemberList(ctx)
-		if err == nil {
-			return g.cachedClient, nil
+		if bootstrapIP, ok := configmap.Annotations[BootstrapIPAnnotationKey]; ok && bootstrapIP != "" {
+			// escape if IPv6
+			if net.ParseIP(bootstrapIP).To4() == nil {
+				bootstrapIP = "[" + bootstrapIP + "]"
+			}
+			etcdEndpoints = append(etcdEndpoints, fmt.Sprintf("https://%s:2379", bootstrapIP))
 		}
+		return etcdEndpoints, nil
 	}
 
-	c, err := getEtcdClientWithClientOpts(etcdEndpoints)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get etcd client: %w", err)
-	}
-	if g.cachedClient != nil {
-		if err := g.cachedClient.Close(); err != nil {
-			utilruntime.HandleError(fmt.Errorf("failed to close cached client: %w", err))
+	newFunc := func() (*clientv3.Client, error) {
+		endpoints, err := endpointFunc()
+		if err != nil {
+			return nil, err
```

Comment:
shall we add more context to the err ?

Reply:
there is more context in both endpointFunc and in the cacheClient. which information do you need here?

Reply:
ok then its fine 👍🏽